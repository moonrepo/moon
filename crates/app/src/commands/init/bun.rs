use super::InitOptions;
use super::prompts::*;
use iocraft::prelude::element;
use miette::IntoDiagnostic;
use moon_config::load_toolchain_bun_config_template;
use moon_console::{
    Console,
    ui::{Container, Entry, Section, Style, StyledText},
};
use moon_pdk_api::{PromptType, SettingPrompt};
use tera::{Context, Tera};
use tracing::instrument;

pub fn render_template(context: Context) -> miette::Result<String> {
    Tera::one_off(load_toolchain_bun_config_template(), &context, false).into_diagnostic()
}

#[instrument(skip_all)]
pub async fn init_bun(console: &Console, options: &InitOptions) -> miette::Result<String> {
    if !options.yes {
        console.render(element! {
            Container {
                Section(title: "Bun") {
                    Entry(
                        name: "Toolchain",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/concepts/toolchain",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                    Entry(
                        name: "Handbook",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/guides/javascript/bun-handbook",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                    Entry(
                        name: "Config",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/config/toolchain#bun",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                }
            }
        })?;
    }

    let bun_version = render_version_prompt(console, options, "Bun", || Ok(None)).await?;

    let sync_dependencies = render_prompt(
        console,
        options,
        &SettingPrompt::new(
            "syncDependencies",
            "Sync project relationships as <file>package.json</file> <property>dependencies</property>?",
            PromptType::Confirm { default: true },
        ),
    )
    .await?;

    let mut context = Context::new();
    if let Some(bun_version) = bun_version {
        context.insert("bun_version", &bun_version);
    }
    context.insert("sync_dependencies", &sync_dependencies);
    context.insert("minimal", &options.minimal);

    render_template(context)
}

#[cfg(test)]
mod tests {
    use super::*;
    use starbase_sandbox::assert_snapshot;

    #[test]
    fn renders_default() {
        let mut context = Context::new();
        context.insert("bun_version", &"1.0.0");
        context.insert("infer_tasks", &false);
        context.insert("sync_dependencies", &true);

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_minimal() {
        let mut context = Context::new();
        context.insert("bun_version", &"1.0.0");
        context.insert("sync_dependencies", &true);
        context.insert("minimal", &true);

        assert_snapshot!(render_template(context).unwrap());
    }
}
