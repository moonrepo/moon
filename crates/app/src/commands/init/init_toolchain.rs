use super::InitOptions;
use super::prompts::prompt_version;
use dialoguer::theme::ColorfulTheme;
use dialoguer::{Confirm, Input, Select};
use miette::IntoDiagnostic;
use moon_console::MoonConsole;
use moon_pdk_api::{
    ConditionType, InitializeToolchainInput, PromptType, SettingCondition, SettingPrompt,
};
use moon_toolchain_plugin::{ToolchainPlugin, ToolchainRegistry};
use schematic::color::apply_style_tags;
use starbase_styles::color;
use starbase_utils::json::JsonValue;
use starbase_utils::yaml::{self, YamlMapping, YamlNumber, YamlValue};
use std::collections::VecDeque;
use std::str::FromStr;
use tracing::instrument;

#[instrument(skip_all)]
pub async fn init_toolchain(
    toolchain_registry: &ToolchainRegistry,
    toolchain: &ToolchainPlugin,
    options: &InitOptions,
    theme: &ColorfulTheme,
    console: &MoonConsole,
    include_locator: bool,
) -> miette::Result<String> {
    // No instructions, so render an empty block
    if !toolchain.has_func("initialize_toolchain").await {
        return Ok(format!("{}: {{}}", toolchain.id));
    }

    // Extract information from the plugin
    let output = toolchain
        .initialize_toolchain(InitializeToolchainInput {
            context: toolchain_registry.create_context(),
        })
        .await?;

    if !options.yes {
        console.print_header(&toolchain.metadata.name)?;

        console.out.write_raw(|buffer| {
            buffer.extend_from_slice(
                format!(
                    "Toolchain: {}\n",
                    color::url("https://moonrepo.dev/docs/concepts/toolchain")
                )
                .as_bytes(),
            );

            if let Some(url) = &output.docs_url {
                buffer.extend_from_slice(format!("Handbook: {}\n", color::url(url)).as_bytes());
            }

            if let Some(url) = &output.config_url {
                buffer.extend_from_slice(format!("Config: {}\n\n", color::url(url)).as_bytes());
            }

            Ok(())
        })?;

        console.out.flush()?;
    }

    // Gather built-in settings
    let mut settings = YamlMapping::new();

    if include_locator {
        settings.insert(
            YamlValue::String("plugin".into()),
            YamlValue::String(toolchain.locator.to_string()),
        );
    }

    if toolchain.has_func("detect_version_files").await {
        if let Some(version) = toolchain.detect_version(&options.dir).await? {
            settings.insert(
                YamlValue::String("version".into()),
                YamlValue::String(version.to_string()),
            );
        }
    }

    if !settings.contains_key("version") && toolchain.supports_tier_3().await {
        // TODO rewrite
        let version = prompt_version(&toolchain.metadata.name, options, theme, || {
            Ok(String::new())
        })?;

        if !version.is_empty() {
            settings.insert(
                YamlValue::String("version".into()),
                YamlValue::String(version),
            );
        }
    }

    // Gather user settings via prompts
    for (key, value) in output.default_settings {
        inject_setting(key, value, &mut settings);
    }

    evaluate_prompts(&output.prompts, &mut settings, options, theme)?;

    // Render into a YAML string
    let config = YamlValue::Mapping(YamlMapping::from_iter([(
        YamlValue::String(toolchain.id.to_string()),
        YamlValue::Mapping(settings),
    )]));

    Ok(yaml::format(&config)?)
}

fn evaluate_prompts(
    prompts: &[SettingPrompt],
    settings: &mut YamlMapping,
    options: &InitOptions,
    theme: &ColorfulTheme,
) -> miette::Result<()> {
    for prompt in prompts
        .iter()
        .filter(|p| if options.minimal { p.minimal } else { true })
    {
        if let Some(condition) = &prompt.condition {
            if !evaluate_condition(condition, settings) {
                continue;
            }
        }

        if let Some(value) = render_prompt(prompt, options, theme)? {
            let falsy = is_json_falsy(&value);

            if prompt.skip_if_falsy && falsy {
                continue;
            }

            inject_setting(prompt.setting.clone(), value, settings);

            if !falsy {
                evaluate_prompts(&prompt.prompts, settings, options, theme)?;
            }
        }
    }

    Ok(())
}

fn evaluate_condition(condition: &SettingCondition, settings: &YamlMapping) -> bool {
    let Some(value) = settings.get(&condition.setting) else {
        return condition.op == ConditionType::NotExists;
    };

    match (&condition.op, value) {
        (ConditionType::BoolEquals(expected), YamlValue::Bool(actual)) => expected == actual,
        (ConditionType::FloatEquals(expected), YamlValue::Number(number)) => {
            number.as_f64().is_some_and(|actual| expected == &actual)
        }
        (ConditionType::IntEquals(expected), YamlValue::Number(number)) => {
            number.as_i64().is_some_and(|actual| expected == &actual)
        }
        (ConditionType::StringContains(needle), YamlValue::String(haystack)) => {
            haystack.contains(needle)
        }
        (ConditionType::StringEquals(expected), YamlValue::String(actual)) => expected == actual,
        _ => false,
    }
}

fn render_prompt(
    prompt: &SettingPrompt,
    options: &InitOptions,
    theme: &ColorfulTheme,
) -> miette::Result<Option<JsonValue>> {
    match &prompt.ty {
        PromptType::None => Ok(None),
        PromptType::Confirm { default } => {
            let result = if options.yes {
                *default
            } else {
                let confirm = Confirm::with_theme(theme)
                    .with_prompt(apply_style_tags(&prompt.question))
                    .default(*default)
                    .show_default(true);

                if prompt.required {
                    confirm.interact().into_diagnostic()?
                } else {
                    confirm
                        .interact_opt()
                        .into_diagnostic()?
                        .unwrap_or(*default)
                }
            };

            Ok(Some(JsonValue::Bool(result)))
        }
        PromptType::Input { default } => {
            let result = if options.yes {
                default.to_owned()
            } else {
                Input::with_theme(theme)
                    .with_prompt(apply_style_tags(&prompt.question))
                    .default(default.to_owned())
                    .show_default(true)
                    .allow_empty(!prompt.required)
                    .interact_text()
                    .into_diagnostic()?
            };

            Ok(Some(JsonValue::String(result)))
        }
        PromptType::Select {
            default_index,
            options: items,
        } => {
            let index = if options.yes {
                *default_index
            } else {
                let labels = items.iter().map(display_json_value).collect::<Vec<_>>();
                let select = Select::with_theme(theme)
                    .with_prompt(apply_style_tags(&prompt.question))
                    .items(&labels)
                    .default(*default_index);

                if prompt.required {
                    select.interact().into_diagnostic()?
                } else {
                    select
                        .interact_opt()
                        .into_diagnostic()?
                        .unwrap_or(*default_index)
                }
            };

            Ok(Some(items[index].clone()))
        }
    }
}

fn inject_setting(key: String, value: JsonValue, settings: &mut YamlMapping) {
    let mut keys = key.split('.').collect::<VecDeque<_>>();

    inject_setting_nested(&mut keys, value, settings);
}

fn inject_setting_nested(keys: &mut VecDeque<&str>, value: JsonValue, settings: &mut YamlMapping) {
    let Some(key) = keys.pop_front() else { return };

    // If no other keys, this is the leaf
    if keys.is_empty() {
        settings.insert(YamlValue::String(key.into()), convert_json_to_yaml(value));
        return;
    }

    // If not an object, change it to one
    if !settings.contains_key(key) || settings.get(key).is_some_and(|inner| !inner.is_mapping()) {
        settings.insert(
            YamlValue::String(key.into()),
            YamlValue::Mapping(YamlMapping::new()),
        );
    }

    // Traverse another depth
    if let Some(YamlValue::Mapping(inner)) = settings.get_mut(key) {
        inject_setting_nested(keys, value, inner);
    }
}

fn convert_json_to_yaml(value: JsonValue) -> YamlValue {
    match value {
        JsonValue::Null => YamlValue::Null,
        JsonValue::Bool(boolean) => YamlValue::Bool(boolean),
        JsonValue::Number(number) => {
            YamlValue::Number(YamlNumber::from_str(&number.to_string()).unwrap())
        }
        JsonValue::String(string) => YamlValue::String(string),
        JsonValue::Array(list) => {
            YamlValue::Sequence(list.into_iter().map(convert_json_to_yaml).collect())
        }
        JsonValue::Object(map) => YamlValue::Mapping(YamlMapping::from_iter(
            map.into_iter()
                .map(|(key, value)| (YamlValue::String(key), convert_json_to_yaml(value))),
        )),
    }
}

fn is_json_falsy(value: &JsonValue) -> bool {
    match value {
        JsonValue::Null => true,
        JsonValue::Bool(boolean) => !(*boolean),
        JsonValue::Number(number) => number.as_f64().is_some_and(|no| no == 0.0),
        JsonValue::String(string) => string.is_empty(),
        JsonValue::Array(list) => list.is_empty(),
        JsonValue::Object(map) => map.is_empty(),
    }
}

fn display_json_value(value: &JsonValue) -> String {
    match value {
        // Remove quotes
        JsonValue::String(string) => string.to_owned(),
        other => other.to_string(),
    }
}
