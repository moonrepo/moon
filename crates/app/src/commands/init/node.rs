use super::InitOptions;
use super::prompts::*;
use iocraft::prelude::element;
use miette::IntoDiagnostic;
use moon_config::load_toolchain_node_config_template;
use moon_console::{
    Console,
    ui::{Container, Entry, Section, Style, StyledText},
};
use moon_lang::{is_using_dependency_manager, is_using_version_manager};
use moon_node_lang::PackageJsonCache;
use moon_pdk_api::{PromptType, SettingPrompt};
use proto_core::UnresolvedVersionSpec;
use starbase_utils::fs;
use starbase_utils::json::JsonValue;
use tera::{Context, Tera};
use tracing::instrument;

pub fn render_template(context: Context) -> miette::Result<String> {
    Tera::one_off(load_toolchain_node_config_template(), &context, false).into_diagnostic()
}

fn detect_node_version(options: &InitOptions) -> miette::Result<Option<UnresolvedVersionSpec>> {
    Ok(if is_using_version_manager(&options.dir, ".nvmrc") {
        UnresolvedVersionSpec::parse(fs::read_file(options.dir.join(".nvmrc"))?).ok()
    } else if is_using_version_manager(&options.dir, ".node-version") {
        UnresolvedVersionSpec::parse(fs::read_file(options.dir.join(".node-version"))?).ok()
    } else {
        None
    })
}

fn detect_node_version_manager(options: &InitOptions) -> miette::Result<String> {
    Ok(if is_using_version_manager(&options.dir, ".nvmrc") {
        "nvm".to_owned()
    } else if is_using_version_manager(&options.dir, ".node-version") {
        "nodenv".to_owned()
    } else {
        String::new()
    })
}

async fn detect_package_manager(
    console: &Console,
    options: &InitOptions,
) -> miette::Result<(String, Option<UnresolvedVersionSpec>)> {
    let mut pm_type = String::new();
    let mut pm_version = String::new();

    // Extract value from `packageManager` field
    if let Ok(Some(pkg)) = PackageJsonCache::read(&options.dir) {
        if let Some(pm) = pkg.data.package_manager {
            if pm.contains('@') {
                let mut parts = pm.split('@');

                pm_type = parts.next().unwrap_or_default().to_owned();
                pm_version = parts.next().unwrap_or_default().to_owned();

                // Remove corepack hash
                if let Some(index) = pm_version.find('+') {
                    pm_version = pm_version[0..index].to_owned();
                }
            } else {
                pm_type = pm;
            }
        }
    }

    // If no value, detect based on files
    if pm_type.is_empty() {
        if is_using_dependency_manager(&options.dir, "yarn.lock") {
            pm_type = "yarn".to_owned();
        } else if is_using_dependency_manager(&options.dir, "pnpm-lock.yaml") {
            pm_type = "pnpm".to_owned();
        } else if is_using_dependency_manager(&options.dir, "bun.lockb") {
            pm_type = "bun".to_owned();
        } else if is_using_dependency_manager(&options.dir, "package-lock.json") {
            pm_type = "npm".to_owned();
        }
    }

    // If no value again, ask for explicit input
    if pm_type.is_empty() {
        let pm = render_prompt(
            console,
            options,
            &SettingPrompt::new(
                "packageManager",
                "Package manager?",
                PromptType::Select {
                    default_index: 0,
                    options: vec!["npm".into(), "pnpm".into(), "yarn".into(), "bun".into()],
                },
            ),
        )
        .await?;

        if let Some(JsonValue::String(inner)) = pm {
            pm_type = inner;
        }
    }

    let pm_version = render_version_prompt(console, options, &pm_type, || {
        if pm_version.is_empty() {
            Ok(None)
        } else {
            Ok(UnresolvedVersionSpec::parse(&pm_version).ok())
        }
    })
    .await?;

    Ok((pm_type, pm_version))
}

#[instrument(skip_all)]
pub async fn init_node(console: &Console, options: &InitOptions) -> miette::Result<String> {
    if !options.yes {
        console.render(element! {
            Container {
                Section(title: "Node") {
                    Entry(
                        name: "Toolchain",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/concepts/toolchain",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                    Entry(
                        name: "Handbook",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/guides/javascript/node-handbook",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                    Entry(
                        name: "Config",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/config/toolchain#node",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                }
            }
        })?;
    }

    let node_version =
        render_version_prompt(console, options, "Node", || detect_node_version(options)).await?;
    let node_version_manager = detect_node_version_manager(options)?;
    let package_manager = detect_package_manager(console, options).await?;

    let infer_tasks = render_prompt(
        console,
        options,
        &SettingPrompt::new(
            "inferTasks",
            "Infer <file>package.json</file> scripts as moon tasks? <muted>(not recommended)</muted>",
            PromptType::Confirm { default: false },
        ),
    )
    .await?;

    let sync_dependencies = render_prompt(
        console,
        options,
        &SettingPrompt::new(
            "syncDependencies",
            "Sync project relationships as <file>package.json</file> <property>dependencies</property>?",
            PromptType::Confirm { default: true },
        ),
    )
    .await?;

    let dedupe_lockfile = render_prompt(
        console,
        options,
        &SettingPrompt::new(
            "dedupeLockfile",
            "Automatically dedupe lockfile when changed?",
            PromptType::Confirm { default: true },
        ),
    )
    .await?;

    let mut context = Context::new();
    if let Some(node_version) = node_version {
        context.insert("node_version", &node_version);
    }
    context.insert("node_version_manager", &node_version_manager);
    context.insert("package_manager", &package_manager.0);
    if let Some(package_manager_version) = package_manager.1 {
        context.insert("package_manager_version", &package_manager_version);
    }
    context.insert("infer_tasks", &infer_tasks);
    context.insert("sync_dependencies", &sync_dependencies);
    context.insert("dedupe_lockfile", &dedupe_lockfile);
    context.insert("minimal", &options.minimal);

    render_template(context)
}

#[cfg(test)]
mod tests {
    use super::*;
    use starbase_sandbox::assert_snapshot;

    fn create_context() -> Context {
        let mut context = Context::new();
        context.insert("node_version", &"16.0.0");
        context.insert("node_version_manager", &"");
        context.insert("package_manager", &"npm");
        context.insert("package_manager_version", &"8.0.0");
        context.insert("infer_tasks", &false);
        context.insert("dedupe_lockfile", &false);
        context.insert("sync_dependencies", &true);
        context.insert("minimal", &false);
        context
    }

    #[test]
    fn renders_default() {
        assert_snapshot!(render_template(create_context()).unwrap());
    }

    #[test]
    fn renders_minimal() {
        let mut context = create_context();
        context.insert("minimal", &true);

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_nvm() {
        let mut context = create_context();
        context.insert("node_version_manager", &"nvm");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_nodenv() {
        let mut context = create_context();
        context.insert("node_version_manager", &"nodenv");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_npm() {
        let mut context = create_context();
        context.insert("package_manager", &"npm");
        context.insert("package_manager_version", &"9.0.0");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_pnpm() {
        let mut context = create_context();
        context.insert("package_manager", &"pnpm");
        context.insert("package_manager_version", &"7.14.0");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_yarn() {
        let mut context = create_context();
        context.insert("package_manager", &"yarn");
        context.insert("package_manager_version", &"3.2.0");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_bun() {
        let mut context = create_context();
        context.insert("package_manager", &"bun");
        context.insert("package_manager_version", &"1.0.0");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_tasks() {
        let mut context = create_context();
        context.insert("infer_tasks", &true);

        assert_snapshot!(render_template(context).unwrap());
    }
}
