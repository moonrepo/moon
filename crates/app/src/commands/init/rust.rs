use super::InitOptions;
use super::prompts::*;
use iocraft::prelude::element;
use miette::IntoDiagnostic;
use moon_config::load_toolchain_rust_config_template;
use moon_console::{
    Console,
    ui::{Container, Entry, Section, Style, StyledText},
};
use moon_rust_lang::toolchain_toml::ToolchainTomlCache;
use proto_core::UnresolvedVersionSpec;
use std::path::Path;
use tera::{Context, Tera};
use tracing::instrument;

fn render_template(context: Context) -> miette::Result<String> {
    Tera::one_off(load_toolchain_rust_config_template(), &context, false).into_diagnostic()
}

fn detect_rust_version(dest_dir: &Path) -> miette::Result<Option<UnresolvedVersionSpec>> {
    if let Some(toolchain_toml) = ToolchainTomlCache::read(dest_dir)? {
        if let Some(version) = toolchain_toml.toolchain.channel {
            let rust_version = if version == "stable"
                || version == "beta"
                || version == "nightly"
                || version.starts_with("nightly")
            {
                Some(version)
            } else {
                fully_qualify_version(version)
            };

            return Ok(rust_version.and_then(|v| UnresolvedVersionSpec::parse(v).ok()));
        }
    }

    Ok(None)
}

#[instrument(skip_all)]
pub async fn init_rust(console: &Console, options: &InitOptions) -> miette::Result<String> {
    if !options.yes {
        console.render(element! {
            Container {
                Section(title: "Rust") {
                    Entry(
                        name: "Toolchain",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/concepts/toolchain",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                    Entry(
                        name: "Handbook",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/guides/rust/handbook",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                    Entry(
                        name: "Config",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/config/toolchain#rust",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                }
            }
        })?;
    }

    let rust_version = render_version_prompt(console, options, "Rust", || {
        detect_rust_version(&options.dir)
    })
    .await?;

    let mut context = Context::new();
    if let Some(rust_version) = rust_version {
        context.insert("rust_version", &rust_version);
    }
    context.insert("minimal", &options.minimal);

    render_template(context)
}

#[cfg(test)]
mod tests {
    use super::*;
    use starbase_sandbox::assert_snapshot;

    #[test]
    fn renders_default() {
        let mut context = Context::new();
        context.insert("rust_version", &"1.70.0");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_minimal() {
        let mut context = Context::new();
        context.insert("rust_version", &"1.70.0");
        context.insert("minimal", &true);

        assert_snapshot!(render_template(context).unwrap());
    }
}
