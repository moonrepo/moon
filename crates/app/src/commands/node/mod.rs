mod run_script;

pub use run_script::{RunScriptArgs, run_script};

use clap::Subcommand;

#[derive(Clone, Debug, Subcommand)]
pub enum NodeCommands {
    #[command(
        name = "run-script",
        about = "Run a `package.json` script within a project."
    )]
    RunScript(RunScriptArgs),
}
