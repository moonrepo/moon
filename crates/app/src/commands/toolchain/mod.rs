pub mod add;
pub mod info;

use clap::Subcommand;

#[derive(Clone, Debug, Subcommand)]
pub enum ToolchainCommands {
    #[command(
        name = "add",
        about = "Add and configure a toolchain plugin in .moon/toolchain.yml."
    )]
    Add(add::ToolchainAddArgs),

    #[command(
        name = "info",
        about = "Show detailed information about a toolchain plugin."
    )]
    Info(info::ToolchainInfoArgs),
}
