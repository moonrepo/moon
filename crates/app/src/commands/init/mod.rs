mod bun;
mod node;
mod prompts;
mod rust;
mod toolchain;

use crate::session::MoonSession;
use bun::init_bun;
use clap::Args;
use clean_path::Clean;
use iocraft::prelude::{FlexDirection, View, element};
use miette::IntoDiagnostic;
use moon_common::{Id, consts::CONFIG_DIRNAME, is_test_env};
use moon_config::{
    ToolchainConfig, load_toolchain_config_template, load_workspace_config_template,
};
use moon_console::{
    Console,
    ui::{Confirm, Container, Notice, StyledText, Variant},
};
use moon_vcs::{Git, Vcs};
use node::init_node;
use proto_core::{Id as PluginId, PluginLocator};
use rust::init_rust;
use starbase::AppResult;
use starbase_styles::color;
use starbase_utils::fs;
use std::collections::BTreeMap;
use std::path::PathBuf;
use tera::{Context, Tera};
use toolchain::init_toolchain;
use tracing::instrument;

#[derive(Args, Clone, Debug)]
pub struct InitArgs {
    #[arg(help = "Specific toolchain to initialize")]
    toolchain: Option<Id>,

    #[arg(help = "Plugin locator for the toolchain")]
    plugin: Option<PluginLocator>,

    #[arg(
        long = "to",
        help = "Destination to initialize into",
        default_value = "."
    )]
    dest: String,

    #[arg(long, help = "Overwrite existing configurations")]
    force: bool,

    #[arg(long, help = "Initialize with minimal configuration and prompts")]
    minimal: bool,

    #[arg(long, help = "Skip prompts and use default values")]
    yes: bool,
}

fn render_toolchain_template(context: &Context) -> miette::Result<String> {
    Tera::one_off(load_toolchain_config_template(), context, false).into_diagnostic()
}

fn render_workspace_template(context: &Context) -> miette::Result<String> {
    Tera::one_off(load_workspace_config_template(), context, false).into_diagnostic()
}

fn create_default_context() -> Context {
    let mut context = Context::new();
    context.insert("projects", &BTreeMap::<String, String>::new());
    context.insert("project_globs", &vec!["apps/*", "packages/*"]);
    context.insert("vcs_manager", &"git");
    context.insert("vcs_provider", &"github");
    context.insert("vcs_default_branch", &"master");
    context
}

fn detect_vcs_provider(repo_root: PathBuf) -> String {
    if repo_root.join(".gitlab").exists() {
        "gitlab".into()
    } else if repo_root.join("bitbucket-pipelines.yml").exists() {
        "bitbucket".into()
    } else {
        "github".into()
    }
}

pub struct InitOptions {
    pub dir: PathBuf,
    pub force: bool,
    pub minimal: bool,
    pub yes: bool,
}

/// Verify the destination and return a path to the `.moon` folder
/// if all questions have passed.
async fn verify_dest_dir(
    console: &Console,
    options: &InitOptions,
) -> miette::Result<Option<PathBuf>> {
    let init = if options.yes {
        true
    } else {
        let mut value = false;

        console
            .render_interactive(element! {
                Confirm(
                    label: format!("Initialize moon into <path>{}</path>?", options.dir.display()),
                    on_confirm: &mut value
                )
            })
            .await?;

        value
    };

    if init {
        let moon_dir = options.dir.join(CONFIG_DIRNAME);

        if !options.force && moon_dir.exists() {
            let mut force = false;

            console
                .render_interactive(element! {
                    Confirm(
                        label: "moon has already been initialized, overwrite it?",
                        on_confirm: &mut force
                    )
                })
                .await?;

            if !force {
                return Ok(None);
            }
        }

        fs::create_dir_all(&moon_dir)?;

        return Ok(Some(moon_dir));
    }

    Ok(None)
}

pub async fn init_for_toolchain(
    session: &MoonSession,
    args: &InitArgs,
    options: &InitOptions,
) -> AppResult {
    let console = &session.console;
    let id = args.toolchain.as_ref().unwrap();

    if !is_test_env() && !options.dir.join(CONFIG_DIRNAME).exists() {
        console.err.write_line(format!(
            "moon has not been initialized! Try running {} first?",
            color::shell("moon init")
        ))?;

        return Ok(Some(1));
    }

    let tool_config = match id.as_str() {
        "bun" => init_bun(console, options).await?,
        "node" => init_node(console, options).await?,
        "rust" => init_rust(console, options).await?,
        _ => {
            let mut include_locator = true;
            let plugin_id = PluginId::raw(id.as_str());
            let plugin_locator = match args.plugin.as_ref() {
                Some(locator) => locator.to_owned(),
                None => match ToolchainConfig::get_plugin_locator(id) {
                    Some(locator) => {
                        include_locator = false;
                        locator
                    }
                    None => {
                        console.err.write_line(
                            "A plugin locator is required as the 2nd argument when initializing a toolchain!"
                        )?;

                        return Ok(Some(1));
                    }
                },
            };

            let toolchain_registry = session.get_toolchain_registry().await?;

            toolchain_registry
                .load_without_config(&plugin_id, plugin_locator)
                .await?;

            let toolchain = toolchain_registry.get_instance(&plugin_id).await?;

            init_toolchain(
                &session.console,
                options,
                &toolchain_registry,
                &toolchain,
                include_locator,
            )
            .await?
        }
    };

    let toolchain_config_path = &session.config_loader.get_toolchain_files(&options.dir)[0];

    if !toolchain_config_path.exists() {
        fs::write_file(
            toolchain_config_path,
            render_toolchain_template(&Context::new())?.trim(),
        )?;
    }

    fs::append_file(toolchain_config_path, format!("\n\n{}", tool_config.trim()))?;

    session.console.render(element! {
        Container {
            Notice(variant: Variant::Success) {
                StyledText(
                    content: "Configuration <file>.moon/toolchain.yml</file> has successfully been updated!"
                )
            }
        }
    })?;

    Ok(None)
}

#[instrument(skip_all)]
pub async fn init(session: MoonSession, args: InitArgs) -> AppResult {
    let dest_path = PathBuf::from(&args.dest);
    let dest_dir = if args.dest == "." {
        session.working_dir.clone()
    } else if dest_path.is_absolute() {
        dest_path
    } else {
        session.working_dir.join(&args.dest)
    };

    let options = InitOptions {
        dir: dest_dir.clean(),
        force: args.force,
        minimal: args.minimal,
        yes: args.yes,
    };

    // Initialize a specific tool and exit early
    if args.toolchain.is_some() {
        init_for_toolchain(&session, &args, &options).await?;

        return Ok(None);
    }

    // Extract template variables
    if verify_dest_dir(&session.console, &options).await?.is_none() {
        return Ok(None);
    }

    let git = Git::load(&options.dir, "master", &[])?;

    let mut context = create_default_context();
    context.insert("vcs_manager", "git");
    context.insert(
        "vcs_provider",
        &detect_vcs_provider(git.get_repository_root().await?),
    );
    context.insert(
        "vcs_default_branch",
        if git.is_enabled() {
            git.get_remote_default_branch().await?
        } else {
            git.get_default_branch().await?
        }
        .as_str(),
    );

    // Create workspace file
    fs::write_file(
        &session.config_loader.get_workspace_files(&options.dir)[0],
        render_workspace_template(&context)?,
    )?;

    // Append to ignore file
    fs::append_file(
        options.dir.join(".gitignore"),
        r#"
# moon
.moon/cache
.moon/docker
"#,
    )?;

    session.console.render(element! {
        Container {
            Notice(variant: Variant::Success) {
                StyledText(
                    content: format!(
                        "Successfully initialized moon in <path>{}</path>!", options.dir.display(),
                    )
                )

                View(padding_top: 1, flex_direction: FlexDirection::Column) {
                    StyledText(content: "Learn more: <url>https://moonrepo.dev/docs</url>")
                    StyledText(content: "Need help?  <url>https://discord.gg/qCh9MEynv2</url>")
                }
            }
        }
    })?;

    Ok(None)
}

#[cfg(test)]
mod tests {
    use super::*;
    use starbase_sandbox::assert_snapshot;

    #[test]
    fn renders_default() {
        let context = create_default_context();

        assert_snapshot!(render_workspace_template(&context).unwrap());
    }

    #[test]
    fn renders_glob_list() {
        let mut context = create_default_context();
        context.insert("project_globs", &vec!["apps/*", "packages/*"]);

        assert_snapshot!(render_workspace_template(&context).unwrap());
    }

    #[test]
    fn renders_projects_map() {
        let mut context = create_default_context();
        context.insert("projects", &BTreeMap::from([("example", "apps/example")]));

        assert_snapshot!(render_workspace_template(&context).unwrap());
    }

    #[test]
    fn renders_git_vcs() {
        let mut context = create_default_context();
        context.insert("vcs_manager", &"git");
        context.insert("vcs_default_branch", &"main");

        assert_snapshot!(render_workspace_template(&context).unwrap());
    }

    #[test]
    fn renders_svn_vcs() {
        let mut context = create_default_context();
        context.insert("vcs_manager", &"svn");
        context.insert("vcs_default_branch", &"trunk");

        assert_snapshot!(render_workspace_template(&context).unwrap());
    }

    #[test]
    fn renders_gitlab() {
        let mut context = create_default_context();
        context.insert("vcs_manager", &"git");
        context.insert("vcs_provider", &"gitlab");
        context.insert("vcs_default_branch", &"main");

        assert_snapshot!(render_workspace_template(&context).unwrap());
    }

    #[test]
    fn renders_bitbucket() {
        let mut context = create_default_context();
        context.insert("vcs_manager", &"git");
        context.insert("vcs_provider", &"bitbucket");
        context.insert("vcs_default_branch", &"main");

        assert_snapshot!(render_workspace_template(&context).unwrap());
    }
}
