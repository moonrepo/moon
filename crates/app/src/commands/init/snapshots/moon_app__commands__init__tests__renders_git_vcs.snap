---
source: crates/app/src/commands/init/mod.rs
expression: render_workspace_template(&context).unwrap()
---
# https://moonrepo.dev/docs/config/workspace
$schema: './cache/schemas/workspace.json'

# extends: './shared/workspace.yml'

projects:
  - 'apps/*'
  - 'packages/*'
vcs:
  manager: 'git'
  defaultBranch: 'main'
